### PR TITLE
Arrow Flight Auth 1 and 2

### LINKED ISSUE #997: integrate with flight auth
We were able to transition to DoExchange and get rid of the BarrageService without actually making Flight's Auth work. We thought that was a good opportunity to split the project off (as the other app_metadata changes have been large enough).

I came across this document describing the intended plan for flight auth the last time it seems to have changed:
https://docs.google.com/document/d/1k3hdbGeE89jiMZ2oKpoIZEs7HAQmVs_RWOmU7fTpLjQ/edit#heading=h.99wjos12oqt

Authorization configuration/enforcement is part of #2563.

### PR DESCRIPTION
**Pull request by nbauernfeind**
We're moving away from SessionService::newSession to Flight Auth. This leaves the javascript side alone, but otherwise enables flight based auth server side.

Fixes #997.

### CODE REVIEW

File: server/src/main/java/io/deephaven/server/arrow/FlightServiceGrpcImpl.java [hunk 1]
```diff
@@ -45,24 +55,102 @@ public class FlightServiceGrpcImpl extends FlightServiceGrpc.FlightServiceImplBa
...
+                    // TODO: anonymous auth via handshake?
+
+                    final WrappedAuthenticationRequest req = WrappedAuthenticationRequest.parseFrom(payload);
+                    final AuthenticationRequestHandler handler = authRequestHandlers.get(req.getType());
+                    if (handler != null) {
+                        auth = handler.login(protocolVersion, req.getPayload(), handshakeResponseListener);
+                    }
+                }
+            } catch (final AuthenticationException | InvalidProtocolBufferException e) {
+                auth = Optional.empty();
```
Comment (niloc132) [inline server/src/main/java/io/deephaven/server/arrow/FlightServiceGrpcImpl.java:128]: this is one of the cases i could reasonably see sending more than just one "it didnt work" error - the contract for empty() from the providers is "no one knew how to handle your request"... eh i could go either way.

probably should log pb errors though?

File: server/src/main/java/io/deephaven/server/arrow/FlightServiceGrpcImpl.java [hunk 2]
```diff
@@ -45,24 +55,102 @@ public class FlightServiceGrpcImpl extends FlightServiceGrpc.FlightServiceImplBa
...
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            // ignore
+        }
+
+        @Override
+        public void onCompleted() {
+            // ignore
```
Comment (niloc132) [inline server/src/main/java/io/deephaven/server/arrow/FlightServiceGrpcImpl.java:152]: probably should close if we never got a request passed to onNext

**Review by niloc132 — COMMENTED** [review]
The main thing I don't see is how to handle multi-step auth, how the server would say "okay here's a nonce to sign" or something using ... errors from the Authorization header or Handshake() call

File: Util/src/main/java/io/deephaven/util/auth/AuthContext.java [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import io.deephaven.util.annotations.FinalDefault;
 import org.jetbrains.annotations.NotNull;
 
+// TODO before merge, move this or split it up and provide some kind of simple mapping for Util etc to use
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply Util/src/main/java/io/deephaven/util/auth/AuthContext.java:9]: We'll move this file to `authentication` package, implement `LogOutputAppendable`, and remove references to `getLogRepresentation`.

File: authentication/src/main/java/io/deephaven/auth/AuthenticationRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+     * Note that regular arrow flight clients cannot specify the protocolVersion; to be compatible with flight auth
+     * assume protocolVersion will be zero.
+     *
+     * @param protocolVersion Mostly unused, this is an allowed field to set on HandshakeRequests from the Flight gRPC
+     *        call.
+     * @param payload The byte payload of the handshake, such as an encoded protobuf.
+     * @param listener The handshake response observer, which enables multi-request authentication.
+     * @return AuthContext for this user if applicable else Empty
+     */
+    Optional<AuthContext> login(long protocolVersion, ByteString payload, HandshakeResponseListener listener)
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply authentication/src/main/java/io/deephaven/auth/AuthenticationRequestHandler.java:44]: Changing to ByteBuffer

**Review by niloc132 — APPROVED** [review]

**Pull request merged**